[visionOS] Refine transition for Image Fullscreen

etiennesegonzac opened the pull request:
#### 078b086740b309270f57e55054d1da8298ffce4f
<pre>
[visionOS] Refine transition for Image Fullscreen
<a href="https://bugs.webkit.org/show_bug.cgi?id=272829">https://bugs.webkit.org/show_bug.cgi?id=272829</a>
&lt;<a href="https://rdar.apple.com/126218680">rdar://126218680</a>&gt;

Reviewed by Aditya Keerthi.

Special case the transition to fullscreen when using Quick Look, since
scene swapping will take care of the animation, and we don&apos;t display the
fullscreen window.
Bump the QOS of the WorkQueue. It&apos;s only used as part of the fullscreen
transition, in reaction to user action.

* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::sharedQuickLookFileQueue):
QOS bump.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
Keep track of whether or not we&apos;re using Image Fullscreen. It impacts
the exit transition, and we don&apos;t want to depend on the exact moment the
image buffer is released in WebFullScreenManagerProxy.
(-[WKFullScreenWindowController _configureSpatialFullScreenTransition]):
Skip the spatial fullscreen transition window manipulation when using
Image Fullscreen.
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
Add a new Image Fullscreen code path for the transition in and out of
fullscreen. Hide the fullscreen window since we&apos;re not using it.
Remove the old branches at different points during the transition.
(-[WKFullScreenWindowController showUI]):
Remove the Image Fullscreen specific chrome management, the WebKit
window will be hidden.

Canonical link: <a href="https://commits.webkit.org/278327@main">https://commits.webkit.org/278327@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b88bcdd6f3808c73584daa34fcbbc85aef61b07f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50171 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29462 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2464 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53429 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/861 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35688 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/471 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40938 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52270 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27137 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43177 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22035 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24560 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8549 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46548 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/489 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55012 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25267 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/433 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48335 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26528 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43362 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47349 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11012 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27391 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26260 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1713,6 +1717,10 @@ - (BOOL)prefersSceneDimming
+#if ENABLE(QUICKLOOK_FULLSCREEN)
+    if (_isImageElement)
+        return;
+#endif // QUICKLOOK_FULLSCREEN
```
Nit: Let's add a newline after the `#endif`.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1789,9 +1797,24 @@ - (void)_performSpatialFullScreenTransition:(BOOL)enter completionHandler:(Compl
+                    completionHandler();
+                });
+            } else {
+                [controller->_fullscreenViewController dismissViewControllerAnimated:NO completion:nil];
```
I think the `controller->` here is unnecessary?

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1789,9 +1797,24 @@ - (void)_performSpatialFullScreenTransition:(BOOL)enter completionHandler:(Compl
-    bool isImageElement = false;
+    auto* manager = self._manager;
+    if (manager) {
+        if (_isImageElement) {
```
These can be merged.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1909,12 +1913,7 @@ - (void)showUI
 {
 #if PLATFORM(VISION)
     UIWindowScene *scene = [_window windowScene];
-#if ENABLE(QUICKLOOK_FULLSCREEN)
```
Did you also want to skip `hideUI` for your use case?

etiennesegonzac committed:
[visionOS] Refine transition for Image Fullscreen
https://bugs.webkit.org/show_bug.cgi?id=272829
<rdar://126218680>

Reviewed by Aditya Keerthi.

Special case the transition to fullscreen when using Quick Look, since
scene swapping will take care of the animation, and we don't display the
fullscreen window.
Bump the QOS of the WorkQueue. It's only used as part of the fullscreen
transition, in reaction to user action.

* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::sharedQuickLookFileQueue):
QOS bump.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
Keep track of whether or not we're using Image Fullscreen. It impacts
the exit transition, and we don't want to depend on the exact moment the
image buffer is released in WebFullScreenManagerProxy.
(-[WKFullScreenWindowController _configureSpatialFullScreenTransition]):
Skip the spatial fullscreen transition window manipulation when using
Image Fullscreen.
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
Add a new Image Fullscreen code path for the transition in and out of
fullscreen. Hide the fullscreen window since we're not using it.
Remove the old branches at different points during the transition.
(-[WKFullScreenWindowController showUI]):
Remove the Image Fullscreen specific chrome management, the WebKit
window will be hidden.

Canonical link: https://commits.webkit.org/278327@main
